### PR TITLE
Implementation of multi DevAddr support

### PR DESCRIPTION
**Pull request by Mandur**
* Multi devaddr support
* Test cases for multi devaddr support
* Added configurable IP for dev template
* Corrected log to hub variable in the ARM template

NOTE: Need a redeployment of the Azure Function code in order to work properly.
fix AB#691

### CODE REVIEW

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageProcessor.cs [hunk 1]
```diff
@@ -51,427 +52,430 @@ public async Task<byte[]> ProcessMessageAsync(byte[] message)
...
+            if (loraDeviceInfoCacheList != null)
             {
-                loraDeviceInfo = await this.loraDeviceInfoManager.GetLoraDeviceInfoAsync(devAddr, this.configuration.GatewayID);
-                if (loraDeviceInfo.DevEUI != null)
-                    Logger.Log(loraDeviceInfo.DevEUI, $"processing message, device not in cache", Logger.LoggingLevel.Info);
-                else
-                    Logger.Log(devAddr, $"processing message, device not in cache", Logger.LoggingLevel.Info);
-                Cache.AddToCache(devAddr, loraDeviceInfo);
+                loraDeviceInfo = loraDeviceInfoCacheList.Values.FirstOrDefault(x => 
+                x.NwkSKey!=null?loraMessage.CheckMic(x.NwkSKey):false==true);
```
Comment (fbeltrao) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageProcessor.cs]: Cosmetic: don't need the `==true`

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageProcessor.cs [hunk 2]
```diff
@@ -486,38 +490,34 @@ private async Task<byte[]> ProcessLoraMessage(LoRaMessageWrapper loraMessage)
 
 
 
-            Logger.Log(loraDeviceInfo.DevEUI, $"processing time: {DateTime.UtcNow - startTimeProcessing}", Logger.LoggingLevel.Info);
+            Logger.Log(loraDeviceInfo!=null?loraDeviceInfo.DevEUI:devAddr, $"processing time: {DateTime.UtcNow - startTimeProcessing}", Logger.LoggingLevel.Info);
```
Comment (fbeltrao) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageProcessor.cs]: Cosmetic: can simplify with `loraDeviceInfo?.DevEUI ?? devAddr`

File: LoRaEngine/test/LoRaWan.IntegrationTest/ABPTest.cs [hunk 3]
```diff
@@ -220,6 +220,79 @@ public async Task Test_ABP_Invalid_NwkSKey_Fails_With_Mic_Error()
...
+            const int MESSAGES_COUNT = 2;
+
+            await SendABPMessages(MESSAGES_COUNT, this.TestFixture.Device16_ABP);
+            await SendABPMessages(MESSAGES_COUNT, this.TestFixture.Device17_ABP);
+
+        }
+
+        private async Task SendABPMessages(int MESSAGES_COUNT, TestDeviceInfo device)
+        {
+            Log($"[INFO] ** Starting {nameof(Test_ABP_Confirmed_And_Unconfirmed_Message)} using device {device.DeviceID} **");
```
Comment (fbeltrao) [inline LoRaEngine/test/LoRaWan.IntegrationTest/ABPTest.cs]: This is the wrong place. Should be where test (with Fact attribute) is.
Log($"[INFO] ** Starting {nameof(Test_ABP_Device_With_Same_DevAddr)} using devices {this.TestFixture.Device16_ABP.DeviceID}, {this.TestFixture.Device17_ABP.DeviceID} **");

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageProcessor.cs [hunk 4]
```diff
@@ -486,38 +490,34 @@ private async Task<byte[]> ProcessLoraMessage(LoRaMessageWrapper loraMessage)
...
 
 
-            Logger.Log(loraDeviceInfo.DevEUI, $"processing time: {DateTime.UtcNow - startTimeProcessing}", Logger.LoggingLevel.Info);
+            Logger.Log(loraDeviceInfo!=null?loraDeviceInfo.DevEUI:devAddr, $"processing time: {DateTime.UtcNow - startTimeProcessing}", Logger.LoggingLevel.Info);
 
             return udpMsgForPktForwarder;
         }
 
-
+        private static void createRequestCache(string devAddr, LoraDeviceInfo loraDeviceInfo)
```
Comment (fbeltrao) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageProcessor.cs]: Any reason to create a method with a single call? All other usages of `Cache` are done directly. Are you going to extend in the future to add more functionality?

Cosmetic: methods start with capital letter

**Comment by ronniesa** [comment]
@fbeltrao any chance to get a pass on the CI? View that is a big change in logic I think it would be nice to have a validation.

**Comment by fbeltrao** [comment]
@ronniesa cannot run the normal CI because the azure pipeline in this PR is older than the one currently available (stuck to old vsts-agent version).

I executed manually and it worked (had to run a few times due to edgeHub issues)

**Comment by ronniesa** [comment]
@fbeltrao I'm fine with this. I will wait for @Mandur changes then but for me the PR is good.

**Comment by Mandur** [comment]
@fbeltrao  Thanks for the review and comments! I changed them accordingly to your feedback!
I reran the tests on the bits and everything went well.

File: LoRaEngine/test/LoRaWan.IntegrationTest/ABPTest.cs [hunk 5]
```diff
@@ -220,6 +220,81 @@ public async Task Test_ABP_Invalid_NwkSKey_Fails_With_Mic_Error()
...
+
+        // Verifies that ABP confirmed and unconfirmed messages are working
+        // Uses Device5_ABP
+        [Fact]
+        public async Task Test_ABP_Device_With_Same_DevAddr()
+        {
+
+
+            const int MESSAGES_COUNT = 2;
+            Log($"[INFO] ** Starting {nameof(Test_ABP_Confirmed_And_Unconfirmed_Message)} using devices {this.TestFixture.Device16_ABP.DeviceID} and {this.TestFixture.Device17_ABP.DeviceID} **");
```
Comment (fbeltrao) [inline LoRaEngine/test/LoRaWan.IntegrationTest/ABPTest.cs]: Correct the name used in log `nameof(Test_ABP_Device_With_Same_DevAddr)`

Comment (Mandur) [reply LoRaEngine/test/LoRaWan.IntegrationTest/ABPTest.cs]: Sorry my bad, forgot this one, It is corrected now

**Comment by ronniesa** [comment]
@fbeltrao can you approve the requested changes so I can merge. thx

**Pull request merged**